src/Features/LanguageServer/Protocol/Handler/DocumentChanges/DidOpenHandler.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.DocumentChanges
+{
+    [Shared]
+    [ExportLspMethod(LSP.Methods.TextDocumentDidOpenName, mutatesSolutionState: true)]
+    internal class DidOpenHandler : IRequestHandler<LSP.DidOpenTextDocumentParams, object>
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public DidOpenHandler()
+        {
+        }
+
+        public LSP.TextDocumentIdentifier? GetTextDocumentIdentifier(LSP.DidOpenTextDocumentParams request)
+            => new LSP.TextDocumentIdentifier { Uri = request.TextDocument.Uri };
+
+        public async Task<object> HandleRequestAsync(LSP.DidOpenTextDocumentParams request, RequestContext context, CancellationToken cancellationToken)
+        {
+            var document = context.Document;
+            Contract.ThrowIfNull(document);
+
+            // Add the document and ensure the text we have matches whats on the client
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            var sourceText = SourceText.From(request.TextDocument.Text, text.Encoding, text.ChecksumAlgorithm);
```

Comment:
Also: in the non-LSP case, it looks like we use the encoding that came from the editor's sniffing and not whatever the prior contents was. This _may_ matter in funky debugging cases.

Reply:
Yeah, this is probably wrong. I think LSP will send everything in UTF-16 (being json serialized) but I didn't know if there were implications of changing the encoding or algorithm.

Reply:
The only use of this encoding and checksum is for debugging purposes; the PDB contains checksums of binary information, not textual information, so encoding is important. I don't know if it matters in the LSP case though at all, since we're not using it for debugging.

Reply:
In that case it seems like encoding and checksum are irrelevant to this case, so I can just remove the `GetTextAsync` call altogether :)